src/util/SE3.py

```diff
@@ -1,102 +1,93 @@
+from geometry_msgs.msg import TransformStamped, Vector3, Quaternion
+from .SO3 import SO3
 
 
-# TODO: adhere to https://github.com/umrover/mrover-ros/discussions/46
-class SE3(Pose):
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
+class SE3:
+    position: np.ndarray
+    rotation: SO3
 
-    @classmethod
-    def from_tf(cls, tf: Transform) -> SE3:
-        """
-        Create an SE3 pose object from a Transform message.
-        The translation of the Transform will become the position of the Pose,
-        and the rotation of the Transform will become the orientation of the Pose.
-
-        :param tf: the Transform message object
-        :returns: the created SE3 pose object
+    def __init__(self, position: np.ndarray = None, rotation: np.ndarray = None):
         """
-        point = vector3_to_point(tf.translation)
-        orientation = tf.rotation
-        return SE3(position=point, orientation=orientation)
+        Initialize an SE3 object
 
-    def to_tf(self) -> Transform:
+        :param position: optional numpy position vector [x, y, z], defaults to zero vector
+        :param rotation: optional numpy quaternion vector [x, y, z, w], defaults to [0, 0, 0, 1]
         """
-        Create a transform message from an SE3 pose object.
-        The opposite of from_tf(), see its docs for details.
+        if position is None:
+            self.position = np.zeros(3)
+        else:
+            self.position = position
 
-        :returns: the created Transform message object
-        """
-        translation = point_to_vector3(self.position)
-        return Transform(translation=translation, rotation=self.orientation)
+        if rotation is None:
+            self.rotation = SO3()
+        else:
+            self.rotation = SO3(rotation)
 
-    def x_vector(self) -> np.ndarray:
-        """
-        Get the unit direction vector of the SE3 pose's X axis.
-
-        :returns: unit direction vector [x, y, z]
-        """
-        rotation_matrix = quaternion_matrix(self.quaternion_array())
-
-        # Extract what the x-axis (forward) is with respect to our rover rotation
-        return rotation_matrix[0:3, 0]
+    @classmethod
+    def from_tf_tree(cls, tf_buffer: tf2_ros.Buffer, parent_frame: str, child_frame: str) -> Optional[SE3]:
+        """
+        Ask the TF tree for a transform from parent_frame to child_frame,
+        and return it as an SE3.
+
+        :param tf_buffer: the tf buffer used to query the TF tree
+        :param parent_frame: the parent frame of the desired transform
+        :param child_frame: the child frame of the desired transform
+
+        :returns: an SE3 containing the tranform from parent_frame to child_frame,
+                  or None if the transform can't be found
+        """
+        try:
+            tf_msg = tf_buffer.lookup_transform(parent_frame, child_frame, rospy.Time()).transform
+
+        except (
```

Comment:
idk I'm kinda torn on this, on one hand I agree that it might be stylistically better to explicitly catch errors higher up rather than checking for None, but on the other hand this saves us from writing a large try except statement every time you look up something in the TF tree (which we do a lot). Its also worth noting that this is a very common exception for us to run into since it takes a short amount of time for TFs to be published and accessible. Also, I feel like returning None actually makes sense and isn't just a convoluted error checker, because nothing has gone catastrophically wrong, you just looked for a TF and it wasn't there so you get nothing.

Reply:
if nothing else, by returning None you've completely obfuscated the underlying error. Not every caller will want to handle a `LookupException`, `ConnectivityException` and `ExtrapolationException` the same way. e.g. Perhaps in some applications a lookup exception is worth a retry, but a connectivity exception needs to be a soft quit.

Reply:
Ok fair enough, ig we just have to deal with a ton of try except blocks. So you think I should just get rid of the try catch block in `from_tf_tree()` and add it back to the places I removed it from the drive code?

Reply:
also should I add some :raises: tags to the docstring for each of these exceptions explaining what they are?

Reply:
yep sg on both

Reply:
Done, now the nav code should be effectively back where it was before this PR. There are a lot of places where these errors aren't caught at all and we need to add try except statements in the nav code, but it's a non trivial addition that will require thinking about it carefully so I'll open an issue for someone to do later